File: Algorithms/SortingSeaching/BinaryInsertionSort.js
"use strict";

let findIndex = (nums, value, leftIndex, rightIndex) => {

    if (rightIndex <= leftIndex) {

        return value > nums[leftIndex] ? leftIndex + 1 : leftIndex;

    }

    let middleIndex = Math.floor((leftIndex + rightIndex) / 2);

    if (value == nums[middleIndex]) {

        return middleIndex + 1;

    } else if (value > nums[middleIndex]) {

        return findIndex(nums, value, middleIndex + 1, rightIndex);

    } else {

        return findIndex(nums, value, leftIndex, middleIndex - 1);

    }

}

let binaryInsertionSort = (nums) => {

    for (let i in nums) {

        let n = nums[i],
            j = i - 1,
            correctIndex = findIndex(nums, n, 0, j);

        while (j >= correctIndex) {

            nums[j + 1] = nums[j];

            j--;

        }

        nums[j + 1] = n;

    }

}

let nums = [1, 2, 0, 8, 22, 189, 221, 10, 88, 67, 45, 34];

console.log(nums);

binaryInsertionSort(nums);

console.log(nums);

File: Algorithms/SortingSeaching/BinaryInsertionSort.py

def _find_index(nums, value, left_index, right_index):

    if right_index <= left_index:

        return left_index + 1 if value > nums[left_index] else left_index

    middle_index = (left_index + right_index) // 2

    if value == nums[middle_index]:

        return middle_index + 1

    elif value > nums[middle_index]:

        return _find_index(nums, value, middle_index + 1, right_index)

    else:

        return _find_index(nums, value, left_index, middle_index - 1)

def binary_insertion_sort(nums):

    for i in range(1, len(nums)):

        n, j = nums[i], i - 1

        correct_index = _find_index(nums, n, 0, j)

        while j >= correct_index:

            nums[j + 1] = nums[j]

            j -= 1

        nums[j + 1] = n

def main():

    nums = [1, 2, 0, 8, 22, 189, 221, 10, 88, 67, 45, 34]

    print(nums)
    binary_insertion_sort(nums)
    print(nums)

if __name__ == '__main__':
    main()
